src/shogun/classifier/Perceptron.cpp

```diff
@@ -78,44 +61,34 @@ bool CPerceptron::train_machine(CFeatures* data)
 		w.set_const(1.0 / num_feat);
 		bias=0;
 	}
-	else
-	{
-		w = get_w();
-	}
-	auto pb = SG_PROGRESS(range(max_iter));
-	//loop till we either get everything classified right or reach max_iter
-	while (!converged && iter < max_iter)
+}
+
+void CPerceptron::iteration()
+{
+	bool converged = true;
+	SGVector<float64_t> w = get_w();
+	SGVector<float64_t> output(m_continue_features->get_num_vectors());
```

Comment:
i think this is not so efficient. @karlnapf ?

Reply:
yes, you are here re-allocating the vector in every iterations. This is not good and we need to find a way to recycle things. As we said we will "communicate" over member variables, I guess this needs to be moved there.